### ISSUE TITLE
🐛 useExhaustiveDependencies incorrectly flags necessary deps as unnecessary in useEffect (v2.3.14)

### ISSUE THREAD

**Issue by PerryFinn** [biomejs/biome#9036]
### Environment information

<details>

```bash

pnpm exec biome rage
CLI:
  Version:                      2.3.14
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.15.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/9.12.3

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    unset

Workspace:
  Open Documents:               0

```

</details>

### What happened?

After upgrading to version 2.3.14, I encountered confusion during the CI pipeline process when trying to upgrade to the new version. After the upgrade, the CI reported numerous errors. I use `Mobx` in my project, but found that the latest modifications don't comply with `Mobx`'s usage guidelines. Personally, I think we need to be cautious about future upgrades.

example:
```tsx
import { observer } from "mobx-react-lite";
import { useEffect } from "react";
import { userStore } from "@/store/userStore";
import { useRouter } from "../hooks";

type Props = {
  children: React.ReactNode;
};
const LoginAuthGuard = ({ children }: Props) => {
  const router = useRouter();
  const accessToken = userStore.userTokenSnapshot;

  useEffect(() => {
    if (!accessToken) {
      router.replace("/auth/login");
    }
// It prompts me to clear the dependency of `accessToken` from Mobx, but I need to track its changes
  }, [router, accessToken]); 

  return <>{children}</>;
};

export default observer(LoginAuthGuard);
```

### Expected result

When using MobX, the `accessToken` variable should be correctly identified as a dependency, just like in ESLint.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Please do a proper search before filing an issue

**Comment by siketyan** [comment]
@ematipico I think this is a separated issue. It complains for the dependency of `accessToken`, which is not used in method access.

Plus it's still reproducible in main branch: https://biomejs.dev/playground/?lintRules=useExhaustiveDependencies&code=aQBtAHAAbwByAHQAIAB7ACAAbwBiAHMAZQByAHYAZQByACAAfQAgAGYAcgBvAG0AIAAiAG0AbwBiAHgALQByAGUAYQBjAHQALQBsAGkAdABlACIAOwAKAGkAbQBwAG8AcgB0ACAAewAgAHUAcwBlAEUAZgBmAGUAYwB0ACAAfQAgAGYAcgBvAG0AIAAiAHIAZQBhAGMAdAAiADsACgBpAG0AcABvAHIAdAAgAHsAIAB1AHMAZQByAFMAdABvAHIAZQAgAH0AIABmAHIAbwBtACAAIgBAAC8AcwB0AG8AcgBlAC8AdQBzAGUAcgBTAHQAbwByAGUAIgA7AAoAaQBtAHAAbwByAHQAIAB7ACAAdQBzAGUAUgBvAHUAdABlAHIAIAB9ACAAZgByAG8AbQAgACIALgAuAC8AaABvAG8AawBzACIAOwAKAAoAdAB5AHAAZQAgAFAAcgBvAHAAcwAgAD0AIAB7AAoAIAAgAGMAaABpAGwAZAByAGUAbgA6ACAAUgBlAGEAYwB0AC4AUgBlAGEAYwB0AE4AbwBkAGUAOwAKAH0AOwAKAGMAbwBuAHMAdAAgAEwAbwBnAGkAbgBBAHUAdABoAEcAdQBhAHIAZAAgAD0AIAAoAHsAIABjAGgAaQBsAGQAcgBlAG4AIAB9ADoAIABQAHIAbwBwAHMAKQAgAD0APgAgAHsACgAgACAAYwBvAG4AcwB0ACAAcgBvAHUAdABlAHIAIAA9ACAAdQBzAGUAUgBvAHUAdABlAHIAKAApADsACgAgACAAYwBvAG4AcwB0ACAAYQBjAGMAZQBzAHMAVABvAGsAZQBuACAAPQAgAHUAcwBlAHIAUwB0AG8AcgBlAC4AdQBzAGUAcgBUAG8AawBlAG4AUwBuAGEAcABzAGgAbwB0ADsACgAKACAAIAB1AHMAZQBFAGYAZgBlAGMAdAAoACgAKQAgAD0APgAgAHsACgAgACAAIAAgAGkAZgAgACgAIQBhAGMAYwBlAHMAcwBUAG8AawBlAG4AKQAgAHsACgAgACAAIAAgACAAIAByAG8AdQB0AGUAcgAuAHIAZQBwAGwAYQBjAGUAKAAiAC8AYQB1AHQAaAAvAGwAbwBnAGkAbgAiACkAOwAKACAAIAAgACAAfQAKAC8ALwAgAEkAdAAgAHAAcgBvAG0AcAB0AHMAIABtAGUAIAB0AG8AIABjAGwAZQBhAHIAIAB0AGgAZQAgAGQAZQBwAGUAbgBkAGUAbgBjAHkAIABvAGYAIABgAGEAYwBjAGUAcwBzAFQAbwBrAGUAbgBgACAAZgByAG8AbQAgAE0AbwBiAHgALAAgAGIAdQB0ACAASQAgAG4AZQBlAGQAIAB0AG8AIAB0AHIAYQBjAGsAIABpAHQAcwAgAGMAaABhAG4AZwBlAHMACgAgACAAfQAsACAAWwByAG8AdQB0AGUAcgAsACAAYQBjAGMAZQBzAHMAVABvAGsAZQBuAF0AKQA7ACAACgAKACAAIAByAGUAdAB1AHIAbgAgADwAPgB7AGMAaABpAGwAZAByAGUAbgB9ADwALwA%2BADsACgB9ADsACgAKAGUAeABwAG8AcgB0ACAAZABlAGYAYQB1AGwAdAAgAG8AYgBzAGUAcgB2AGUAcgAoAEwAbwBnAGkAbgBBAHUAdABoAEcAdQBhAHIAZAApADsA

**Comment by ematipico** [comment]
Apologies!

**Comment by mdevils** [comment]
@PerryFinn could you please point me to the `Mobx`'s usage guidelines where this approach is encouraged?

**Comment by PerryFinn** [comment]
Sure

`Mobx` is a state management library, similar to `zustand`. My team is using `Mobx` for state management.

Since there are many ways to use `Mobx`, here I will use the simplest demo to show how our team uses `Mobx`.

First, we will create a `demo-store-v1.ts` file in a `store` folder to manage the relevant global states I need:

```tsx
// demo-store-v1.ts
import { makeAutoObservable } from "mobx";

export class DemoStore {
  name = "";
  age = 0;

  constructor() {
    makeAutoObservable(
      this,
      { autoBind: true },
    );
  }

  setName(value: string) {
    this.name = value;
  }

  setAge(value: number) {
    this.age = value;
  }
}

export const demoStore = new DemoStore();
```

Sometimes, we also slightly adjust the way of creating a `store` according to different versions of `Mobx`, using decorators:

```tsx
// demo-store-v2.ts
import { action, makeObservable, observable } from 'mobx';

export class DemoStore {
  @observable
  name = '';

  @observable
  age = 0;

  constructor() {
    makeObservable(this); // Use `makeObservable` instead of `makeAutoObservable`.
  }

  @action
  setName(value: string) {
    this.name = value;
  }
  @action
  setAge(value: number) {
    this.age = value;
  }
}

export const demoStore = new DemoStore();
```

However, I believe the two different creation methods of `demo-store-v1.ts` and `demo-store-v2.ts` are not very relevant to this issue, so I will use the `demo-store-v1.ts` method to create the `store`.

After creation, we can use it directly in the component, for example:

```tsx
// demo-profile.tsx
import { demoStore } from './demo-store-v1';
import { observer } from "mobx-react-lite";

function DemoProfile() {
  const { name } = demoStore;
  const age = demoStore.age;
  return (
    <div>
      <h1>Demo Profile</h1>
      <p>Name: {name}</p>
      <p>Age: {age}</p>
    </div>
  );
}

// As long as you use the `observer` package component, the values accessed from the `demoStore` will be reactive, similar to React state
export default observer(DemoProfile);
```

That's all for now. For more different usage methods, we can refer to this guide: https://mobx.js.org/react-integration.html

**Comment by mdevils** [comment]
@PerryFinn Thank you for the detailed response!

However, I believe the rule is actually correct here and is pointing out a real issue in your code. Let me explain.

In your example:

```js
  const accessToken = userStore.userTokenSnapshot;
  useEffect(() => {
    if (!accessToken) {
      router.replace("/auth/login");
    }
  }, [router, accessToken]);
```

`accessToken` comes from a global MobX store (`userStore`), not from component state or props. The `useEffect` dependency array is designed to track values tied to the component's lifecycle (state, props, derived values). A global store value like `userStore.userTokenSnapshot` is not connected to the component lifecycle — it can change at any time independently of renders.

Including `accessToken` in the dependency array doesn't mean "run the effect when accessToken changes." It means "run the effect when accessToken changes and the component re-renders." These are two potentially unrelated moments — the effect won't fire until something else triggers a re-render, which may or may not coincide with the store update.

I also checked the https://mobx.js.org/react-integration.html you linked, and I don't see this pattern being recommended there. The guide recommends wrapping components with `observer()` to make the render reactive to observable changes, and using https://mobx.js.org/reactions.html to perform side effects in response to observable changes — not `useEffect` with store values as dependencies.

In fact, the guide itself contains an example of exactly this situation under the "useEffect and observables" tip — reacting to an observable value inside `useEffect` — and it solves it using `autorun()`:

```js
  useEffect(
      () =>
          autorun(() => {
              if (timer.secondsPassed > 60)
                  alert("Still there. It's a minute already?!!")
          }),
      []
  )
```

For your use case, a similar approach using autorun or reaction would be correct:

```js
  import { reaction } from "mobx";

  useEffect(() => {
    return reaction(
      () => userStore.userTokenSnapshot,
      (accessToken) => {
        if (!accessToken) {
          router.replace("/auth/login");
        }
      },
      { fireImmediately: true }
    );
  }, [router]);
```

This way the side effect fires reliably whenever the observable changes, regardless of React's render cycle.

**Comment by PerryFinn** [comment]
Thank you so much for pointing out the issues in my code. There are indeed such monitoring mechanisms as `autorun` and `reaction` in `Mobx`. Both of your approaches are very correct, and I fully agree with them.

Perhaps it's because my mental model of `React` isn't quite accurate yet. I might tend to think that `useEffect` acts like a monitor, triggering a callback immediately when the value of certain variables changes to perform certain actions.

**Comment by mdevils** [comment]
@PerryFinn you are welcome!
